## Verificación y Validación del Dataset Crediticio

### Introducción
**Contexto**

El presente notebook tiene como objetivo realizar una verificación estructural y validación técnica del dataset crediticio previamente limpiado (credit_dataset_clean.csv).

Antes de construir modelos predictivos o realizar análisis inferenciales, es fundamental garantizar que los datos:

- No contengan valores imposibles o inconsistentes.
- Cumplan con coherencia económica.
- No presenten duplicidades estructurales.
- Mantengan integridad en variables numéricas y categóricas.

**Objetivo del análisis**

Verificar la consistencia estadística, estructural y económica del dataset final, asegurando su idoneidad para modelado crediticio.

In [1]:
import numpy as np
import pandas as pd

In [2]:
#Leer credit_dataset_clean.csv
df_model = pd.read_csv("credit_dataset_clean.csv")

### Descripción de Variables

Este dataset contiene información financiera y de comportamiento crediticio a nivel mensual por cliente. Cada fila representa un cliente en un mes específico.

#### Identificadores

Estas variables permiten identificar de manera única cada observación.

**Customer_ID**

- Tipo: Identificador único
- Descripción: Código único asignado a cada cliente.
- Función: Permite agrupar observaciones por individuo.
- Uso en modelo: No se utiliza como variable predictora.

**Month**

- Tipo: Variable temporal
- Descripción: Mes correspondiente al registro financiero.
- Función: Permite análisis longitudinal por cliente.
- Uso en modelo: Puede utilizarse para análisis de panel o tendencias.


#### Variable Objetivo

**Credit_Score**

Clasificación crediticia original:

- Poor → Perfil de alto riesgo
- Standard → Perfil intermedio
- Good → Perfil sólido

En modelos tradicionales puede transformarse en variable binaria (ej. Poor vs No Poor).

### **Variables Financieras**

A continuación se describen formalmente las variables numéricas incluidas en el dataset.

#### **Variables Demográficas**

**Age**

- Tipo: Numérica discreta
- Descripción: Edad del cliente.
- Interpretación económica: Puede capturar madurez financiera.

Edad del cliente (18–56 años). Se limpió eliminando valores imposibles y usando la edad consistente por cliente.

#### **Variables de Ingreso**

**Annual_Income**

- Tipo: Numérica continua
- Descripción: Ingreso anual bruto del cliente.
- Importancia: Determina capacidad de pago estructural.

Se aplicó winsorización para eliminar valores extremos irreales.

**Monthly_Inhand_Salary**

Tipo: Numérica continua
Descripción: Ingreso mensual disponible después de deducciones.

Cuando estaba ausente se imputó como:

$$MonthlyInhandSalary = \frac{AnnualIncome}{12}$$

#### **Variables de Endeudamiento**

**Outstanding_Debt**

- Tipo: Numérica continua
- Descripción: Deuda total pendiente.
- Interpretación: Nivel de apalancamiento actual.

**Num_of_Loan**

- Tipo: Numérica discreta
- Descripción: Número total de préstamos activos.
- Interpretación: Intensidad de exposición crediticia.

**Total_EMI_per_month**

- Tipo: Numérica continua
- Descripción: Pago mensual total por créditos (EMI).
- Interpretación: Carga financiera mensual.

#### **Variables de Historial Crediticio**

**Credit_History_Months**

- Tipo: Numérica discreta
- Descripción: Antigüedad crediticia en meses.
- Interpretación: Experiencia en el sistema financiero.

**Num_of_Delayed_Payment**

- Tipo: Numérica discreta
- Descripción: Número de pagos retrasados.
- Interpretación: Indicador directo de riesgo.

**Delay_from_due_date**

- Tipo: Numérica discreta
- Descripción: Días promedio de retraso.
- Interpretación: Severidad de mora.

**Num_Credit_Inquiries**

- Tipo: Numérica discreta
- Descripción: Número de consultas crediticias recientes.
- Interpretación: Señal de posible búsqueda activa de crédito.

#### **Variables de Comportamiento Financiero**

**Credit_Utilization_Ratio**

- Tipo: Numérica continua
- Descripción: Proporción del crédito utilizado respecto al disponible.
- Interpretación: Indicador clave de estrés financiero.

**Monthly_Balance**

- Tipo: Numérica continua
- Descripción: Saldo promedio mensual restante después de pagos.
- Interpretación: Liquidez neta.

**Amount_invested_monthly**

Tipo: Numérica continua
Descripción: Monto invertido mensualmente por el cliente.
Interpretación: Capacidad de ahorro/inversión.

#### **Variables de Infraestructura Financiera**

**Num_Bank_Accounts**

- Tipo: Numérica discreta
- Descripción: Número de cuentas bancarias activas.

**Num_Credit_Card**

- Tipo: Numérica discreta
- Descripción: Número de tarjetas de crédito activas.

#### **Variables de Condiciones Contractuales**

**Interest_Rate**

- Tipo: Numérica continua
- Descripción: Tasa de interés promedio aplicada.

**Changed_Credit_Limit**

- Tipo: Numérica continua
- Descripción: Variación en el límite de crédito.
- Interpretación: Ajustes de confianza crediticia.

### **Variables Categóricas Transformadas para Modelado**

Estas variables fueron codificadas ordinalmente para permitir su uso en modelos estadísticos.

**Credit_Mix_final**

Calidad de mezcla crediticia:

| Valor | Significado                   |
| ----- | ----------------------------- |
| 0     | Bad (mezcla crediticia débil) |
| 1     | Standard (mezcla promedio)    |
| 2     | Good (mezcla saludable)       |

Interpretación: Diversificación del portafolio de crédito.

**Payment_of_Min_Amount_final**

| Valor | Significado       |
| ----- | ----------------- |
| 0     | No paga el mínimo |
| 1     | Sí paga el mínimo |

Indicador de disciplina financiera.

**Spending_Level_final**

| Valor | Significado       |
| ----- | ----------------- |
| 0     | Low (gasto bajo)  |
| 1     | High (gasto alto) |

Clasificación del patrón de gasto.

**Payment_Size_final**

| Valor | Significado             |
| ----- | ----------------------- |
| 0     | Small (pagos pequeños)  |
| 1     | Medium (pagos medianos) |
| 2     | Large (pagos grandes)   |

Tamaño promedio de pagos realizados.

### **Variable Auxiliar**

**Occupation**

- Ocupación principal del cliente.
- Se limpió eliminando valores corruptos y asignando la ocupación consistente por cliente.
- No se incluye directamente en el modelo base tradicional, pero es útil para análisis posteriores.


In [3]:
#Verificación de valores imposibles (numéricas)
num_cols = [
    "Age",
    "Annual_Income",
    "Monthly_Inhand_Salary",
    "Outstanding_Debt",
    "Credit_History_Months",
    "Monthly_Balance",
    "Amount_invested_monthly",
    "Num_of_Loan",
    "Num_of_Delayed_Payment",
    "Changed_Credit_Limit",
    "Interest_Rate",
    "Num_Credit_Inquiries",
    "Total_EMI_per_month",
    "Num_Bank_Accounts",
    "Num_Credit_Card",
    "Delay_from_due_date",
    "Credit_Utilization_Ratio"
]

for col in num_cols:
    print("\n", col)
    print("Min:", df_model[col].min())
    print("Max:", df_model[col].max())
    print("Negativos %:", (df_model[col] < 0).mean())
    print("Inf %:", np.isinf(df_model[col]).mean())


 Age
Min: 18.0
Max: 56.0
Negativos %: 0.0
Inf %: 0.0

 Annual_Income
Min: 7005.93
Max: 21752126.01800004
Negativos %: 0.0
Inf %: 0.0

 Monthly_Inhand_Salary
Min: 303.6454166666666
Max: 1812677.16816667
Negativos %: 0.0
Inf %: 0.0

 Outstanding_Debt
Min: 0.23
Max: 4998.07
Negativos %: 0.0
Inf %: 0.0

 Credit_History_Months
Min: 1.0
Max: 404.0
Negativos %: 0.0
Inf %: 0.0

 Monthly_Balance
Min: 0.0077596647753352
Max: 1602.0405189622518
Negativos %: 0.0
Inf %: 0.0

 Amount_invested_monthly
Min: 0.0
Max: 1977.326102249349
Negativos %: 0.0
Inf %: 0.0

 Num_of_Loan
Min: 0.0
Max: 9.0
Negativos %: 0.0
Inf %: 0.0

 Num_of_Delayed_Payment
Min: 0.0
Max: 28.0
Negativos %: 0.0
Inf %: 0.0

 Changed_Credit_Limit
Min: -6.49
Max: 36.97
Negativos %: 0.01586
Inf %: 0.0

 Interest_Rate
Min: 1.0
Max: 34.0
Negativos %: 0.0
Inf %: 0.0

 Num_Credit_Inquiries
Min: 0.0
Max: 17.0
Negativos %: 0.0
Inf %: 0.0

 Total_EMI_per_month
Min: 0.0
Max: 437.01275293262665
Negativos %: 0.0
Inf %: 0.0

 Num_Bank_Accounts
Mi

In [4]:
#Verificación de duplicados estructurales
print("Duplicados exactos:",
      df_model.duplicated().mean())

print("Duplicados Customer_ID + Month:",
      df_model.duplicated(["Customer_ID", "Month"]).mean())

Duplicados exactos: 0.0
Duplicados Customer_ID + Month: 0.0


In [5]:
#Verificación de cardinalidad categórica
cat_cols = [
    "Credit_Score",
    "Occupation"
]

for col in cat_cols:
    print("\n", col)
    print("Unique:", df_model[col].nunique())
    print(df_model[col].value_counts().head(10))


 Credit_Score
Unique: 3
Credit_Score
Standard    53174
Poor        28998
Good        17828
Name: count, dtype: int64

 Occupation
Unique: 15
Occupation
Lawyer           7096
Engineer         6864
Architect        6824
Mechanic         6776
Scientist        6744
Accountant       6744
Media_Manager    6720
Developer        6720
Teacher          6672
Entrepreneur     6648
Name: count, dtype: int64


In [ ]:
#Verificación de rangos económicos razonables
print("Debt > Income %:",
      (df_model["Outstanding_Debt"] > df_model["Annual_Income"]).mean())

print("EMI > Monthly Salary %:",
      (df_model["Total_EMI_per_month"] > df_model["Monthly_Inhand_Salary"]).mean())

print("Invested > Monthly Salary %:",
      (df_model["Amount_invested_monthly"] > df_model["Monthly_Inhand_Salary"]).mean())

In [7]:
#Verificación de correlación extrema
corr = df_model[num_cols].corr()

high_corr = (
    corr.abs()
        .where(np.triu(np.ones(corr.shape), k=1).astype(bool))
        .stack()
        .sort_values(ascending=False)
)

print(high_corr.head(10))

Outstanding_Debt        Num_of_Loan              0.638713
Interest_Rate           Num_Credit_Inquiries     0.638544
Outstanding_Debt        Interest_Rate            0.629414
                        Credit_History_Months    0.629312
Credit_History_Months   Num_Credit_Inquiries     0.617524
Num_of_Delayed_Payment  Num_Bank_Accounts        0.612242
Credit_History_Months   Num_of_Loan              0.605728
Outstanding_Debt        Num_Credit_Inquiries     0.603256
Interest_Rate           Delay_from_due_date      0.585941
                        Num_Bank_Accounts        0.584456
dtype: float64


### Resultados de la Verificación del Dataset

**Integridad Numérica**

Se evaluaron mínimos, máximos, valores negativos y presencia de infinitos en todas las variables numéricas.

Resultados:

- No se detectaron valores infinitos.
- No existen valores negativos en variables donde no son económicamente válidos.
= Los rangos observados son consistentes con comportamiento financiero real.

Caso particular:

Changed_Credit_Limit presenta 1.586% de valores negativos, lo cual es válido ya que representa reducciones en el límite de crédito.

En general, no se identifican anomalías estructurales relevantes.

**Verificación de Duplicados**

Duplicados exactos: 0.0%

Duplicados por Customer_ID + Month: 0.0%

El dataset mantiene integridad estructural y no presenta registros repetidos.

**Variable Objetivo**

Credit_Score presenta 3 categorías:

- Standard: 53,174
- Poor: 28,998
- Good: 17,828

Existe desbalance moderado, siendo Standard la clase mayoritaria. Esto deberá considerarse en etapas de modelado.

**Coherencia Económica**

Se analizaron posibles inconsistencias financieras:

- Debt > Income: 0.0%
- EMI > Monthly Salary: 0.008%
- Investment > Monthly Salary: 0.0%

No se observan incoherencias significativas. Los pocos casos extremos no comprometen la calidad del dataset.

**Correlaciones Relevantes**

Las correlaciones más altas se encuentran alrededor de 0.60–0.64, principalmente entre:

- Deuda y número de préstamos
- Tasa de interés y consultas crediticias

Estas relaciones son económicamente coherentes y no indican multicolinealidad severa.

**Conclusión**

El dataset presenta:

- Consistencia numérica
- Integridad estructural
- Coherencia económica
- Correlaciones razonables

Por lo tanto, se considera apto para modelado de riesgo crediticio.